
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://CultureBotAI.github.io/MicroGrowAgents/cofactor_data_sources/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>Cofactor Data Sources and KG-Microbe Integration - MicroGrowAgents</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cofactor-data-sources-and-kg-microbe-integration" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="MicroGrowAgents" class="md-header__button md-logo" aria-label="MicroGrowAgents" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MicroGrowAgents
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Cofactor Data Sources and KG-Microbe Integration
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/CultureBotAI/MicroGrowAgents" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="MicroGrowAgents" class="md-nav__button md-logo" aria-label="MicroGrowAgents" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    MicroGrowAgents
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/CultureBotAI/MicroGrowAgents" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Overview
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cofactor-analysis-pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      
        Cofactor Analysis Pipeline
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#primary-data-sources" class="md-nav__link">
    <span class="md-ellipsis">
      
        Primary Data Sources
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Primary Data Sources">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#supporting-databases" class="md-nav__link">
    <span class="md-ellipsis">
      
        Supporting Databases
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#specialized-literature" class="md-nav__link">
    <span class="md-ellipsis">
      
        Specialized Literature
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Reference Files
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Reference Files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-cofactor_hierarchyyaml" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. cofactor_hierarchy.yaml
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-ec_to_cofactor_mapyaml" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. ec_to_cofactor_map.yaml
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-ingredient_cofactor_mappingcsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. ingredient_cofactor_mapping.csv
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kg-microbe-integration" class="md-nav__link">
    <span class="md-ellipsis">
      
        KG-Microbe Integration
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="KG-Microbe Integration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#what-kg-microbe-is-used-for" class="md-nav__link">
    <span class="md-ellipsis">
      
        What KG-Microbe IS Used For
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#what-kg-microbe-is-not-used-for" class="md-nav__link">
    <span class="md-ellipsis">
      
        What KG-Microbe is NOT Used For
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kgreasoningagent-query-patterns" class="md-nav__link">
    <span class="md-ellipsis">
      
        KGReasoningAgent Query Patterns
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chebi-ontology-access" class="md-nav__link">
    <span class="md-ellipsis">
      
        ChEBI Ontology Access
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-source-evidence-integration" class="md-nav__link">
    <span class="md-ellipsis">
      
        Multi-Source Evidence Integration
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-flow-in-cofactormediaagent" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Flow in CofactorMediaAgent
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#coverage-metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Coverage Metrics
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citation-summary" class="md-nav__link">
    <span class="md-ellipsis">
      
        Citation Summary
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Citation Summary">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#core-databases-required" class="md-nav__link">
    <span class="md-ellipsis">
      
        Core Databases (Required)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#supporting-databases_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Supporting Databases
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#specialized-literature_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        Specialized Literature
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#related-documentation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Related Documentation
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#version-history" class="md-nav__link">
    <span class="md-ellipsis">
      
        Version History
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="cofactor-data-sources-and-kg-microbe-integration">Cofactor Data Sources and KG-Microbe Integration</h1>
<p>This document provides comprehensive documentation of the datasets, resources, and citations that inform the CofactorMediaAgent's cofactor analysis pipeline.</p>
<h2 id="overview">Overview</h2>
<p>The CofactorMediaAgent integrates <strong>6 major biological databases</strong> plus specialized literature to analyze cofactor requirements for microbial organisms. The system combines:</p>
<ol>
<li><strong>Curated reference data</strong> (cofactor hierarchies, EC mappings)</li>
<li><strong>Organism genome data</strong> (Bakta annotations)</li>
<li><strong>Knowledge graph queries</strong> (KG-Microbe via KGReasoningAgent)</li>
<li><strong>Literature evidence</strong> (organism-specific mechanisms)</li>
</ol>
<h2 id="cofactor-analysis-pipeline">Cofactor Analysis Pipeline</h2>
<pre><code>Bakta GFF3 Genome Annotation
         ↓
Extract EC Numbers (from genome_annotations table)
         ↓
Map EC → Cofactors (ec_to_cofactor_map.yaml)
         ↓
Enrich with Details (cofactor_hierarchy.yaml)
         ↓
Query KG-Microbe for Evidence (KGReasoningAgent)
         ├→ Enzyme-substrate relationships
         ├→ Biosynthesis pathway context
         └→ Acquisition mechanisms (transporters)
         ↓
Map to MP Ingredients (ingredient_cofactor_mapping.csv)
         ↓
Generate Hierarchical Table (CofactorMediaAgent._build_cofactor_table)
</code></pre>
<h2 id="primary-data-sources">Primary Data Sources</h2>
<table>
<thead>
<tr>
<th>Resource</th>
<th>Coverage</th>
<th>Primary Use</th>
<th>DOI</th>
<th>Integration Point</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>ChEBI</strong></td>
<td>44 cofactors</td>
<td>Chemical IDs, names, synonyms</td>
<td><a href="https://doi.org/10.1093/nar/gkv1031">10.1093/nar/gkv1031</a></td>
<td><code>cofactor_hierarchy.yaml</code> IDs</td>
</tr>
<tr>
<td><strong>KEGG</strong></td>
<td>30+ pathways</td>
<td>Biosynthesis pathway definitions</td>
<td><a href="https://doi.org/10.1093/nar/gkac963">10.1093/nar/gkac963</a></td>
<td><code>cofactor_hierarchy.yaml</code> pathways</td>
</tr>
<tr>
<td><strong>EC Nomenclature</strong></td>
<td>68 EC patterns</td>
<td>Enzyme classification</td>
<td><a href="https://doi.org/10.1093/nar/gkn582">10.1093/nar/gkn582</a></td>
<td><code>ec_to_cofactor_map.yaml</code></td>
</tr>
<tr>
<td><strong>BRENDA</strong></td>
<td>68 EC mappings</td>
<td>EC-cofactor relationships</td>
<td><a href="https://doi.org/10.1093/nar/gky1048">10.1093/nar/gky1048</a></td>
<td><code>ec_to_cofactor_map.yaml</code></td>
</tr>
<tr>
<td><strong>KG-Microbe</strong></td>
<td>1.5M nodes, 5.1M edges</td>
<td>Enzyme-substrate queries</td>
<td>N/A (project data)</td>
<td><code>KGReasoningAgent</code> runtime</td>
</tr>
<tr>
<td><strong>MP Medium CSV</strong></td>
<td>158 ingredients</td>
<td>Ingredient chemical properties</td>
<td>Various (90.5% coverage)</td>
<td><code>ingredient_cofactor_mapping.csv</code></td>
</tr>
</tbody>
</table>
<h3 id="supporting-databases">Supporting Databases</h3>
<ul>
<li><strong>MetaCyc</strong>: Pathway-cofactor relationship validation (<a href="https://doi.org/10.1093/nar/gkz862">DOI: 10.1093/nar/gkz862</a>)</li>
<li><strong>UniProt</strong>: Enzyme examples and annotations (<a href="https://doi.org/10.1093/nar/gkac1052">DOI: 10.1093/nar/gkac1052</a>)</li>
<li><strong>Reactome</strong>: Cofactor reaction roles (<a href="https://doi.org/10.1093/nar/gkab1028">DOI: 10.1093/nar/gkab1028</a>)</li>
</ul>
<h3 id="specialized-literature">Specialized Literature</h3>
<ul>
<li><strong>Lanthanide cofactors</strong>: Good NM, et al. (2019) <em>J Biol Chem.</em> 294(21):8437-8446 (<a href="https://doi.org/10.1074/jbc.RA119.008197">DOI: 10.1074/jbc.RA119.008197</a>)</li>
<li><strong>PQQ</strong>: Anthony C. (2001) <em>Antioxid Redox Signal.</em> 3(5):757-774 (<a href="https://doi.org/10.1089/15230860152664966">DOI: 10.1089/15230860152664966</a>)</li>
<li><strong>Methanogen cofactors</strong>: Thauer RK, et al. (2008) <em>Nat Rev Microbiol.</em> 6(8):579-591 (<a href="https://doi.org/10.1038/nrmicro1931">DOI: 10.1038/nrmicro1931</a>)</li>
</ul>
<h2 id="reference-files">Reference Files</h2>
<h3 id="1-cofactor_hierarchyyaml">1. cofactor_hierarchy.yaml</h3>
<p><strong>Location</strong>: <code>src/microgrowagents/data/cofactor_hierarchy.yaml</code></p>
<p><strong>Content</strong>: 44 distinct cofactors organized into 5 functional categories:
- <strong>Vitamins</strong> (14 cofactors): TPP, Biotin, Cobalamin, PLP, FMN, FAD, NAD/NADP, Folate, THF, Menaquinone, Ubiquinone, Lipoic acid
- <strong>Metals</strong> (13 cofactors): Fe-S clusters, Heme (a/b/c/d1), Siroheme, Fe, Zn, Cu, Mg, Ca, Mn, Mo, Co, Ni, W, Lanthanides
- <strong>Nucleotides</strong> (11 cofactors): NAD/NADH, NADP/NADPH, ATP, ADP, GTP, CTP, UTP
- <strong>Energy Cofactors</strong> (7 cofactors): CoA, Acetyl-CoA, SAM, PAPS, cAMP, cGMP
- <strong>Other Specialized</strong> (9 cofactors): PQQ, Methanofuran, H4MPT, CoM, CoB, Biotin CCCP, Retinal, Mycofactocin, F420, Methanophenazine</p>
<p><strong>Data Sources</strong>:
- ChEBI IDs for all 44 cofactors
- KEGG pathway IDs (30+ biosynthesis pathways: ko00730, ko00760, ko00780, etc.)
- EC number associations (100+ enzyme-cofactor relationships)
- Enzyme examples from UniProt
- Specialized literature for rare cofactors</p>
<p><strong>Structure</strong>:</p>
<pre><code class="language-yaml">cofactor_hierarchy:
  vitamins:
    cofactors:
      thiamine_pyrophosphate:
        id: &quot;CHEBI:9532&quot;
        names: [&quot;TPP&quot;, &quot;ThDP&quot;]
        precursor: &quot;thiamine&quot;
        ec_associations: [&quot;2.2.1.-&quot;, &quot;4.1.1.-&quot;]
        kegg_pathways: [&quot;ko00730&quot;]
</code></pre>
<h3 id="2-ec_to_cofactor_mapyaml">2. ec_to_cofactor_map.yaml</h3>
<p><strong>Location</strong>: <code>src/microgrowagents/data/ec_to_cofactor_map.yaml</code></p>
<p><strong>Content</strong>: 68 EC number patterns mapped to required and optional cofactors</p>
<p><strong>Data Sources</strong>:
- BRENDA: Primary EC-to-cofactor relationship data
- ExplorEnz: EC number classification and validation
- MetaCyc: Pathway-based cofactor validation</p>
<p><strong>Pattern Matching</strong>:
- Supports wildcards (e.g., <code>"1.1.1.-"</code> matches all EC 1.1.1.X enzymes)
- Exact match takes precedence over pattern match
- Pattern hierarchy: 4-level &gt; 3-level &gt; 2-level &gt; 1-level</p>
<p><strong>Example</strong>:</p>
<pre><code class="language-yaml">ec_cofactor_mapping:
  &quot;1.1.1.-&quot;: {primary: [&quot;nad&quot;, &quot;nadp&quot;], optional: [&quot;zinc_ion&quot;]}
  &quot;2.6.1.-&quot;: {primary: [&quot;pyridoxal_phosphate&quot;]}
</code></pre>
<h3 id="3-ingredient_cofactor_mappingcsv">3. ingredient_cofactor_mapping.csv</h3>
<p><strong>Location</strong>: <code>data/processed/ingredient_cofactor_mapping.csv</code></p>
<p><strong>Content</strong>: 13 MP medium ingredients identified as cofactor providers</p>
<p><strong>Generation Method</strong>:
- <strong>NOT generated from KG-Microbe</strong>
- Source: <code>data/raw/mp_medium_ingredient_properties.csv</code> (158 ingredients)
- Method: Rule-based pattern matching on component names
- Script: <code>scripts/generate_ingredient_cofactor_mapping.py</code></p>
<p><strong>Example Mappings</strong>:</p>
<pre><code>MgCl₂·6H₂O → magnesium_ion (CHEBI:18420)
FeSO₄·7H₂O → iron_ion, heme, iron_sulfur_clusters
Thiamin → thiamine_pyrophosphate (CHEBI:9532)
CoCl₂·6H₂O → cobalamin, cobalt_ion
</code></pre>
<h2 id="kg-microbe-integration">KG-Microbe Integration</h2>
<h3 id="what-kg-microbe-is-used-for">What KG-Microbe IS Used For</h3>
<p>KG-Microbe (1.5M nodes, 5.1M edges) provides <strong>runtime evidence integration</strong> via the <code>KGReasoningAgent</code>:</p>
<ol>
<li><strong>Enzyme-Substrate Relationship Queries</strong></li>
<li>SQL query pattern (USE CASE 3):
     <code>sql
     SELECT enzyme_id, enzyme_name, relationship
     FROM kg_edges e
     JOIN kg_nodes n ON e.subject = n.id
     WHERE e.object = ? (substrate CHEBI ID)
     AND e.predicate IN ('biolink:has_input', 'biolink:has_participant')
     AND n.category = 'biolink:MolecularActivity'</code></li>
<li>
<p>Finds all enzymes using a specific cofactor substrate</p>
</li>
<li>
<p><strong>Biosynthesis Pathway Context</strong></p>
</li>
<li>Query pathway completeness for cofactor biosynthesis capability</li>
<li>
<p>Determines if organism can produce cofactor de novo</p>
</li>
<li>
<p><strong>Acquisition Mechanisms</strong></p>
</li>
<li>Transporter detection: Genome annotations + KG relationships</li>
<li>Identifies uptake systems for external cofactors</li>
<li>
<p>Detects secreted chelators and binding molecules</p>
</li>
<li>
<p><strong>Multi-Source Evidence Integration</strong></p>
</li>
<li>Combines genome, KG, literature, and KEGG data</li>
<li>Confidence scoring based on evidence source count</li>
<li>File: <code>src/microgrowagents/agents/cofactor_media_agent.py:183-237</code></li>
</ol>
<h3 id="what-kg-microbe-is-not-used-for">What KG-Microbe is NOT Used For</h3>
<ol>
<li><strong>Cofactor Hierarchy Definition</strong></li>
<li>Manually curated from ChEBI, KEGG, BRENDA databases</li>
<li>
<p>NOT derived from KG-Microbe relationships</p>
</li>
<li>
<p><strong>EC-to-Cofactor Mapping</strong></p>
</li>
<li>BRENDA-based curation with ExplorEnz validation</li>
<li>
<p>Independent of KG-Microbe</p>
</li>
<li>
<p><strong>Ingredient Mapping Generation</strong></p>
</li>
<li>Rule-based parsing of MP medium CSV</li>
<li>Pattern matching on component names</li>
<li>NO KG-Microbe queries involved</li>
</ol>
<h3 id="kgreasoningagent-query-patterns">KGReasoningAgent Query Patterns</h3>
<p><strong>File</strong>: <code>src/microgrowagents/agents/kg_reasoning_agent.py</code></p>
<p><strong>Query capabilities</strong>:
- <strong>USE CASE 1</strong>: Organism → Media → Ingredients pathway
- <strong>USE CASE 2</strong>: Phenotype → Media recommendations
- <strong>USE CASE 3</strong>: Enzyme-substrate relationships (cofactor analysis)</p>
<p><strong>Database</strong>: DuckDB-loaded KG data (<code>data/raw/kg_microbe_core/merged-kg_*.tsv</code>)</p>
<p><strong>Graph algorithms</strong>: GRAPE (Graph Representation leArning) for 10-100x performance</p>
<h2 id="chebi-ontology-access">ChEBI Ontology Access</h2>
<p><strong>Direct OWL file parsing</strong> (not through KG-Microbe):
- <strong>Client</strong>: <code>src/microgrowagents/chemistry/api_clients/chebi_client.py</code>
- <strong>Features</strong>: Exact + fuzzy matching, synonym support, normalization
- <strong>Source</strong>: <a href="https://ftp.ebi.ac.uk/pub/databases/chebi/ontology/chebi.owl">https://ftp.ebi.ac.uk/pub/databases/chebi/ontology/chebi.owl</a>
- <strong>Used for</strong>: Ingredient name → ChEBI ID matching</p>
<h2 id="multi-source-evidence-integration">Multi-Source Evidence Integration</h2>
<p>The <code>CofactorMediaAgent.analyze_cofactor_requirements()</code> method integrates:</p>
<ol>
<li><strong>Genome</strong> (Bakta GFF3 EC numbers)</li>
<li>Direct SQL query: <code>genome_annotations</code> table</li>
<li>
<p>Method: <code>_get_ec_numbers_from_genome()</code></p>
</li>
<li>
<p><strong>KG-Microbe</strong> (pathway context)</p>
</li>
<li>Via <code>KGReasoningAgent</code> runtime queries</li>
<li>Enzyme-substrate relationships</li>
<li>
<p>Biosynthesis pathway validation</p>
</li>
<li>
<p><strong>KEGG</strong> (biosynthesis completeness)</p>
</li>
<li>Reference data from <code>cofactor_hierarchy.yaml</code></li>
<li>
<p>Pathway IDs: ko00730, ko00760, etc.</p>
</li>
<li>
<p><strong>Literature</strong> (organism mechanisms)</p>
</li>
<li>Via <code>LiteratureAgent</code> searches</li>
<li>Organism-specific evidence</li>
</ol>
<p><strong>Confidence scoring</strong>:
- High: 3-4 evidence sources agree
- Medium: 2 evidence sources
- Low: 1 evidence source</p>
<h2 id="data-flow-in-cofactormediaagent">Data Flow in CofactorMediaAgent</h2>
<pre><code class="language-python"># Step 1: Extract EC numbers from genome
ec_numbers = self._get_ec_numbers_from_genome(organism)
# Query: SELECT DISTINCT ec_numbers FROM genome_annotations WHERE genome_id = ?

# Step 2: Map EC → Cofactors
for ec in ec_numbers:
    cofactors = self._map_ec_to_cofactors(ec)
    # Uses ec_to_cofactor_map.yaml with pattern matching

# Step 3: Enrich with cofactor details
for cofactor_key in cofactors:
    info = self._get_cofactor_info(cofactor_key)
    # Loads from cofactor_hierarchy.yaml

# Step 4: Query KG-Microbe (optional, for evidence)
# Via KGReasoningAgent for enzyme-substrate relationships

# Step 5: Map to MP ingredients
ingredient_mapping = self.map_ingredients_to_cofactors(requirements, &quot;MP&quot;)
# Uses ingredient_cofactor_mapping.csv

# Step 6: Build hierarchical output table
table = self._build_cofactor_table(requirements, ingredient_mapping)
</code></pre>
<h2 id="coverage-metrics">Coverage Metrics</h2>
<ul>
<li><strong>Cofactors defined</strong>: 44 across 5 categories</li>
<li><strong>KEGG pathways referenced</strong>: 30+</li>
<li><strong>EC number associations</strong>: 100+</li>
<li><strong>EC pattern mappings</strong>: 68</li>
<li><strong>MP medium ingredients mapped</strong>: 13/158 (8.2% are cofactor providers)</li>
<li><strong>Pattern coverage</strong>: ~80% of common microbial enzymes</li>
</ul>
<h2 id="citation-summary">Citation Summary</h2>
<h3 id="core-databases-required">Core Databases (Required)</h3>
<ol>
<li>ChEBI: <a href="https://doi.org/10.1093/nar/gkv1031">10.1093/nar/gkv1031</a></li>
<li>KEGG: <a href="https://doi.org/10.1093/nar/gkac963">10.1093/nar/gkac963</a></li>
<li>BRENDA: <a href="https://doi.org/10.1093/nar/gky1048">10.1093/nar/gky1048</a></li>
<li>ExplorEnz: <a href="https://doi.org/10.1093/nar/gkn582">10.1093/nar/gkn582</a></li>
</ol>
<h3 id="supporting-databases_1">Supporting Databases</h3>
<ol>
<li>MetaCyc: <a href="https://doi.org/10.1093/nar/gkz862">10.1093/nar/gkz862</a></li>
<li>UniProt: <a href="https://doi.org/10.1093/nar/gkac1052">10.1093/nar/gkac1052</a></li>
<li>Reactome: <a href="https://doi.org/10.1093/nar/gkab1028">10.1093/nar/gkab1028</a></li>
</ol>
<h3 id="specialized-literature_1">Specialized Literature</h3>
<ol>
<li>Lanthanides: <a href="https://doi.org/10.1074/jbc.RA119.008197">10.1074/jbc.RA119.008197</a></li>
<li>PQQ: <a href="https://doi.org/10.1089/15230860152664966">10.1089/15230860152664966</a></li>
<li>Methanogens: <a href="https://doi.org/10.1038/nrmicro1931">10.1038/nrmicro1931</a></li>
</ol>
<p>For BibTeX format, see: <a href="../references/cofactor_sources.bib"><code>docs/references/cofactor_sources.bib</code></a></p>
<h2 id="related-documentation">Related Documentation</h2>
<ul>
<li><a href="../cofactor_curation_methodology/">Cofactor Curation Methodology</a> - Detailed curation workflow</li>
<li><a href="../src/microgrowagents/agents/cofactor_media_agent.py">CofactorMediaAgent API</a> - Implementation</li>
<li><a href="../src/microgrowagents/agents/kg_reasoning_agent.py">KGReasoningAgent</a> - KG-Microbe integration</li>
<li><a href="../data/raw/mp_medium_ingredient_properties.csv">MP Medium Dataset</a> - Source ingredient data</li>
</ul>
<h2 id="version-history">Version History</h2>
<ul>
<li><strong>v1.0</strong> (2025-01-10): Initial documentation with comprehensive citations and KG-Microbe integration details</li>
</ul>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tabs.link"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>