
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://CultureBotAI.github.io/MicroGrowAgents/MP_media_optimization_plan/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.1">
    
    
      
        <title>MP Medium Optimization for M. extorquens AM-1 Growth and Neodymium Depletion - MicroGrowAgents</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.484c7ddc.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#mp-medium-optimization-for-m-extorquens-am-1-growth-and-neodymium-depletion" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="MicroGrowAgents" class="md-header__button md-logo" aria-label="MicroGrowAgents" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MicroGrowAgents
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              MP Medium Optimization for M. extorquens AM-1 Growth and Neodymium Depletion
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/CultureBotAI/MicroGrowAgents" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="MicroGrowAgents" class="md-nav__button md-logo" aria-label="MicroGrowAgents" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    MicroGrowAgents
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/CultureBotAI/MicroGrowAgents" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Overview
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#background-why-m-extorquens-am-1-is-ideal-for-nd-depletion" class="md-nav__link">
    <span class="md-ellipsis">
      
        Background: Why M. extorquens AM-1 is Ideal for Nd Depletion
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-by-step-instructions" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step-by-Step Instructions
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step-by-Step Instructions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#phase-1-query-organism-specific-requirements" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 1: Query Organism-Specific Requirements
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Phase 1: Query Organism-Specific Requirements">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-11-detect-auxotrophies" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 1.1: Detect Auxotrophies
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-12-query-methanol-dehydrogenase-enzymes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 1.2: Query Methanol Dehydrogenase Enzymes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-13-query-cofactor-requirements" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 1.3: Query Cofactor Requirements
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phase-2-query-lanthanide-specific-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 2: Query Lanthanide-Specific Data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Phase 2: Query Lanthanide-Specific Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-21-query-neodymium-chemical-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 2.1: Query Neodymium Chemical Data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-22-query-xoxf-gene-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 2.2: Query XoxF Gene Data
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-23-search-lanthanide-metabolism-publications" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 2.3: Search Lanthanide Metabolism Publications
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phase-3-optimize-mp-medium-formulation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 3: Optimize MP Medium Formulation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Phase 3: Optimize MP Medium Formulation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#option-a-optimize-existing-mp-medium-recommended" class="md-nav__link">
    <span class="md-ellipsis">
      
        Option A: Optimize Existing MP Medium (Recommended)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#option-b-design-new-medium-from-scratch" class="md-nav__link">
    <span class="md-ellipsis">
      
        Option B: Design New Medium from Scratch
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#option-c-direct-agent-query-most-flexible" class="md-nav__link">
    <span class="md-ellipsis">
      
        Option C: Direct Agent Query (Most Flexible)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phase-4-analyze-concentration-recommendations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 4: Analyze Concentration Recommendations
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phase-5-interpret-and-verify-results" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 5: Interpret and Verify Results
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Phase 5: Interpret and Verify Results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-51-review-ingredient-roles" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 5.1: Review Ingredient Roles
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-52-validate-against-literature" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 5.2: Validate Against Literature
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phase-6-export-and-use-formulation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 6: Export and Use Formulation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Phase 6: Export and Use Formulation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-61-export-to-standard-format" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 6.1: Export to Standard Format
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-62-generate-preparation-protocol" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 6.2: Generate Preparation Protocol
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#key-considerations-for-nd-depletion" class="md-nav__link">
    <span class="md-ellipsis">
      
        Key Considerations for Nd Depletion
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Key Considerations for Nd Depletion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-calcium-lanthanide-competition" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Calcium-Lanthanide Competition
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-methylolanthanin-production" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Methylolanthanin Production
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-ph-and-solubility" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. pH and Solubility
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-biosorption-vs-bioaccumulation" class="md-nav__link">
    <span class="md-ellipsis">
      
        4. Biosorption vs Bioaccumulation
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#expected-results" class="md-nav__link">
    <span class="md-ellipsis">
      
        Expected Results
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Expected Results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#optimized-mp-medium-formulation-example" class="md-nav__link">
    <span class="md-ellipsis">
      
        Optimized MP Medium Formulation (Example)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#performance-metrics-to-track" class="md-nav__link">
    <span class="md-ellipsis">
      
        Performance Metrics to Track
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#troubleshooting" class="md-nav__link">
    <span class="md-ellipsis">
      
        Troubleshooting
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Troubleshooting">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#poor-growth-od600-05-after-48h" class="md-nav__link">
    <span class="md-ellipsis">
      
        Poor Growth (&lt;OD600 0.5 after 48h)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#low-nd-depletion-50-after-48h" class="md-nav__link">
    <span class="md-ellipsis">
      
        Low Nd Depletion (&lt;50% after 48h)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#nd-precipitation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Nd Precipitation
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cli-alternative" class="md-nav__link">
    <span class="md-ellipsis">
      
        CLI Alternative
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#critical-files-referenced" class="md-nav__link">
    <span class="md-ellipsis">
      
        Critical Files Referenced
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Critical Files Referenced">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#agents" class="md-nav__link">
    <span class="md-ellipsis">
      
        Agents:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#skills" class="md-nav__link">
    <span class="md-ellipsis">
      
        Skills:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#data-files" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Files:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#database" class="md-nav__link">
    <span class="md-ellipsis">
      
        Database:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#verification" class="md-nav__link">
    <span class="md-ellipsis">
      
        Verification
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary" class="md-nav__link">
    <span class="md-ellipsis">
      
        Summary
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="mp-medium-optimization-for-m-extorquens-am-1-growth-and-neodymium-depletion">MP Medium Optimization for M. extorquens AM-1 Growth and Neodymium Depletion</h1>
<h2 id="overview">Overview</h2>
<p>This plan provides step-by-step instructions for using the MicroGrowAgents framework to optimize MP medium for:
- <strong>Primary goal</strong>: Supporting Methylobacterium extorquens AM-1 growth
- <strong>Secondary goal</strong>: Enabling Neodymium (Nd³⁺) depletion from solution</p>
<p>The framework has extensive capabilities for this task including:
- Lanthanide-specific data (XoxF enzyme, methylolanthanin lanthanophore, Nd optimal concentrations)
- Genome-informed optimization for M. extorquens
- Multi-agent workflows for media formulation
- Biosorption process data relevant to Nd uptake</p>
<h2 id="background-why-m-extorquens-am-1-is-ideal-for-nd-depletion">Background: Why M. extorquens AM-1 is Ideal for Nd Depletion</h2>
<p>From the sheets_cmm data exploration:</p>
<ol>
<li>
<p><strong>XoxF enzyme system</strong>: M. extorquens produces lanthanide-dependent methanol dehydrogenase (XoxF, KEGG:K23995) that preferentially binds Nd³⁺ and other lanthanides over Ca²⁺</p>
</li>
<li>
<p><strong>Methylolanthanin production</strong>: This organism produces a lanthanophore (hydroxamate-type siderophore) with high affinity for La/Ce/Nd, enhancing uptake</p>
</li>
<li>
<p><strong>Optimal Nd concentration</strong>: Data shows 2 μM Nd³⁺ is optimal for growth (from ingredient_properties_20260108.tsv)</p>
</li>
<li>
<p><strong>Biosorption capability</strong>: Related process data (BP-001) shows M. extorquens achieves 45 mg Eu³⁺/g biomass (lanthanides behave similarly)</p>
</li>
<li>
<p><strong>Methylotrophic metabolism</strong>: Growth on methanol provides simple carbon source, reducing medium complexity</p>
</li>
</ol>
<h2 id="step-by-step-instructions">Step-by-Step Instructions</h2>
<h3 id="phase-1-query-organism-specific-requirements">Phase 1: Query Organism-Specific Requirements</h3>
<p>Use GenomeFunctionAgent to understand M. extorquens AM-1's genetic capabilities and requirements.</p>
<h4 id="step-11-detect-auxotrophies">Step 1.1: Detect Auxotrophies</h4>
<pre><code class="language-python">from microgrowagents.agents.genome_function_agent import GenomeFunctionAgent
from pathlib import Path

# Initialize agent with database
db_path = Path(&quot;data/processed/microgrow.duckdb&quot;)
genome_agent = GenomeFunctionAgent(db_path)

# Detect what M. extorquens cannot synthesize
auxotrophy_result = genome_agent.detect_auxotrophies(
    query=&quot;detect auxotrophies for Methylobacterium extorquens AM-1&quot;,
    organism=&quot;Methylobacterium extorquens&quot;
)

print(auxotrophy_result)
</code></pre>
<p><strong>Expected output</strong>: Essential amino acids, vitamins, or cofactors that must be supplemented</p>
<h4 id="step-12-query-methanol-dehydrogenase-enzymes">Step 1.2: Query Methanol Dehydrogenase Enzymes</h4>
<pre><code class="language-python"># Find methanol dehydrogenases (XoxF vs MxaF)
enzyme_result = genome_agent.find_enzymes(
    query=&quot;find methanol dehydrogenase enzymes in M. extorquens&quot;,
    organism=&quot;Methylobacterium extorquens&quot;,
    ec_number=&quot;1.1.2.7&quot;  # Methanol dehydrogenase
)

print(f&quot;Found {len(enzyme_result['data']['enzymes'])} methanol dehydrogenases&quot;)
</code></pre>
<p><strong>Key distinction</strong>:
- <strong>XoxF</strong> (lanthanide-dependent): Requires Nd³⁺/La³⁺
- <strong>MxaF</strong> (calcium-dependent): Requires Ca²⁺</p>
<p>For Nd depletion, you want to favor XoxF expression.</p>
<h4 id="step-13-query-cofactor-requirements">Step 1.3: Query Cofactor Requirements</h4>
<pre><code class="language-python"># Find all cofactor requirements
cofactor_result = genome_agent.find_cofactor_requirements(
    query=&quot;find cofactor requirements for M. extorquens methanol metabolism&quot;,
    organism=&quot;Methylobacterium extorquens&quot;
)

print(cofactor_result)
</code></pre>
<p><strong>Expected cofactors</strong>:
- PQQ (pyrroloquinoline quinone) - essential for methanol dehydrogenases
- Lanthanides (Nd³⁺, La³⁺, Ce³⁺) - for XoxF
- B vitamins (B12 for C1 metabolism)</p>
<h3 id="phase-2-query-lanthanide-specific-data">Phase 2: Query Lanthanide-Specific Data</h3>
<p>Use SheetQuerySkill to access extensive lanthanide metabolism data from sheets_cmm.</p>
<h4 id="step-21-query-neodymium-chemical-data">Step 2.1: Query Neodymium Chemical Data</h4>
<pre><code class="language-python">from microgrowagents.skills.simple.sheet_query import SheetQuerySkill

# Initialize skill
sheet_skill = SheetQuerySkill(db_path)

# Look up Neodymium compound data
nd_result = sheet_skill.run(
    collection_id=&quot;cmm&quot;,
    query_type=&quot;entity_lookup&quot;,
    entity_name=&quot;neodymium&quot;,
    output_format=&quot;evidence_report&quot;
)

print(nd_result)
</code></pre>
<p><strong>What you'll find</strong>:
- Chemical properties of Nd³⁺
- Optimal concentration ranges (2 μM for growth)
- Solubility and precipitation concerns
- Compatible buffer systems</p>
<h4 id="step-22-query-xoxf-gene-data">Step 2.2: Query XoxF Gene Data</h4>
<pre><code class="language-python"># Cross-reference XoxF with related entities
xoxf_result = sheet_skill.run(
    collection_id=&quot;cmm&quot;,
    query_type=&quot;cross_reference&quot;,
    source_entity_id=&quot;K23995&quot;,  # XoxF KEGG ID
    output_format=&quot;evidence_report&quot;
)

print(xoxf_result)
</code></pre>
<p><strong>What you'll find</strong>:
- XoxF enzyme properties
- Cofactor binding (Nd³⁺ vs Ca²⁺ affinity)
- Related pathways (methanol oxidation, formaldehyde assimilation)
- Publications with experimental conditions</p>
<h4 id="step-23-search-lanthanide-metabolism-publications">Step 2.3: Search Lanthanide Metabolism Publications</h4>
<pre><code class="language-python"># Full-text search in 120+ publications
pub_result = sheet_skill.run(
    collection_id=&quot;cmm&quot;,
    query_type=&quot;publication_search&quot;,
    keyword=&quot;neodymium methylobacterium&quot;,
    output_format=&quot;markdown&quot;
)

print(pub_result)
</code></pre>
<p><strong>What you'll find</strong>:
- Growth curves with different lanthanides
- Optimal Nd concentrations
- Synergistic effects with other media components
- Biosorption/bioaccumulation data</p>
<h3 id="phase-3-optimize-mp-medium-formulation">Phase 3: Optimize MP Medium Formulation</h3>
<p>Use OptimizeMediumWorkflow or MediaFormulationAgent to generate optimized formulation.</p>
<h4 id="option-a-optimize-existing-mp-medium-recommended">Option A: Optimize Existing MP Medium (Recommended)</h4>
<pre><code class="language-python">from microgrowagents.skills.workflows.optimize_medium import OptimizeMediumWorkflow

# Initialize workflow
workflow = OptimizeMediumWorkflow(db_path)

# Optimize MP medium for M. extorquens with Nd
result = workflow.run(
    medium_name=&quot;MP medium&quot;,
    organism=&quot;Methylobacterium extorquens AM-1&quot;,
    optimization_goal=&quot;growth&quot;,  # Prioritize growth (Nd depletion follows)
    additional_context=&quot;&quot;&quot;
    Optimize for:
    1. Lanthanide-dependent growth (favor XoxF over MxaF)
    2. Neodymium biosorption/bioaccumulation
    3. Methylolanthanin production (lanthanophore for Nd uptake)

    Known optimal conditions:
    - 2 μM Nd³⁺ (neodymium chloride)
    - 0.5-1% methanol as carbon source
    - pH 7.0, 30°C, aerobic
    - PQQ supplementation (0.5-1 mg/L) if needed
    &quot;&quot;&quot;
)

print(result)
</code></pre>
<p><strong>Expected output</strong>:
- Complete optimized medium formulation with concentrations
- Rationale for each ingredient
- Cost analysis
- Predicted growth rate</p>
<h4 id="option-b-design-new-medium-from-scratch">Option B: Design New Medium from Scratch</h4>
<pre><code class="language-python">from microgrowagents.skills.workflows.recommend_media import RecommendMediaWorkflow

workflow = RecommendMediaWorkflow(db_path)

result = workflow.run(
    query=&quot;Design defined medium for M. extorquens with neodymium depletion&quot;,
    organism=&quot;Methylobacterium extorquens AM-1&quot;,
    growth_conditions={
        &quot;temperature&quot;: 30,
        &quot;pH&quot;: 7.0,
        &quot;oxygen&quot;: &quot;aerobic&quot;,
        &quot;carbon_source&quot;: &quot;methanol&quot;
    },
    formulation_goals=[
        &quot;defined&quot;,  # Chemically defined, no complex ingredients
        &quot;lanthanide_optimized&quot;,  # Favor XoxF pathway
        &quot;cost_effective&quot;
    ],
    special_requirements=[
        &quot;2 μM neodymium chloride&quot;,
        &quot;PQQ supplementation&quot;,
        &quot;methylolanthanin production support&quot;
    ]
)

print(result)
</code></pre>
<h4 id="option-c-direct-agent-query-most-flexible">Option C: Direct Agent Query (Most Flexible)</h4>
<pre><code class="language-python">from microgrowagents.agents.media_formulation_agent import MediaFormulationAgent

agent = MediaFormulationAgent(db_path)

result = agent.run(
    query=&quot;&quot;&quot;
    Optimize MP medium for Methylobacterium extorquens AM-1 to achieve:
    1. Maximum growth rate
    2. Maximum neodymium depletion from solution

    Starting conditions:
    - Standard MP medium base
    - Initial Nd concentration: 50 μM (adjust based on your system)
    - Target final Nd: &lt;5 μM after 48h

    Consider:
    - XoxF vs MxaF expression (favor XoxF)
    - Methylolanthanin production
    - Calcium-lanthanide competition
    - PQQ availability
    &quot;&quot;&quot;,
    organism=&quot;Methylobacterium extorquens AM-1&quot;,
    base_medium=&quot;MP medium&quot;
)

print(result)
</code></pre>
<h3 id="phase-4-analyze-concentration-recommendations">Phase 4: Analyze Concentration Recommendations</h3>
<p>Use GenMediaConcAgent to predict optimal concentration ranges.</p>
<pre><code class="language-python">from microgrowagents.agents.gen_media_conc_agent import GenMediaConcAgent

conc_agent = GenMediaConcAgent(db_path)

# Get Nd³⁺ concentration recommendation
nd_conc = conc_agent.run(
    query=&quot;optimal neodymium concentration for Methylobacterium extorquens growth&quot;,
    organism=&quot;Methylobacterium extorquens&quot;,
    ingredient=&quot;neodymium chloride&quot;,
    role=&quot;RARE_EARTH_ELEMENT&quot;
)

print(f&quot;Recommended Nd³⁺ range: {nd_conc}&quot;)
</code></pre>
<h3 id="phase-5-interpret-and-verify-results">Phase 5: Interpret and Verify Results</h3>
<h4 id="step-51-review-ingredient-roles">Step 5.1: Review Ingredient Roles</h4>
<pre><code class="language-python">from microgrowagents.agents.media_role_agent import MediaRoleAgent

role_agent = MediaRoleAgent(db_path)

# Verify all ingredients have correct roles
for ingredient in optimized_medium['ingredients']:
    role_result = role_agent.run(
        query=f&quot;classify role of {ingredient['name']}&quot;,
        ingredient_name=ingredient['name'],
        context=&quot;methylotrophic growth with lanthanide cofactors&quot;
    )
    print(f&quot;{ingredient['name']}: {role_result['data']['role']}&quot;)
</code></pre>
<p><strong>Expected roles</strong>:
- Methanol → CARBON_SOURCE
- Nd³⁺ → RARE_EARTH_ELEMENT (cofactor)
- PQQ → COFACTOR
- Methylolanthanin precursors → CHELATOR/SIDEROPHORE</p>
<h4 id="step-52-validate-against-literature">Step 5.2: Validate Against Literature</h4>
<pre><code class="language-python"># Cross-check recommendations against publications
validation_result = sheet_skill.run(
    collection_id=&quot;cmm&quot;,
    query_type=&quot;publication_search&quot;,
    keyword=&quot;methylobacterium extorquens neodymium growth&quot;,
    output_format=&quot;evidence_report&quot;
)

# Compare recommended concentrations with published data
print(validation_result)
</code></pre>
<h3 id="phase-6-export-and-use-formulation">Phase 6: Export and Use Formulation</h3>
<h4 id="step-61-export-to-standard-format">Step 6.1: Export to Standard Format</h4>
<pre><code class="language-python"># Export optimized medium to TSV
import pandas as pd

ingredients_df = pd.DataFrame(result['data']['formulation']['ingredients'])
ingredients_df.to_csv('optimized_MP_medium_for_Nd_depletion.tsv',
                      sep='\t', index=False)

print(&quot;Exported to: optimized_MP_medium_for_Nd_depletion.tsv&quot;)
</code></pre>
<h4 id="step-62-generate-preparation-protocol">Step 6.2: Generate Preparation Protocol</h4>
<pre><code class="language-python"># Get step-by-step preparation instructions
protocol = agent.generate_protocol(
    formulation=result['data']['formulation'],
    volume=1000,  # mL
    include_sterilization=True
)

print(protocol)
</code></pre>
<h2 id="key-considerations-for-nd-depletion">Key Considerations for Nd Depletion</h2>
<h3 id="1-calcium-lanthanide-competition">1. Calcium-Lanthanide Competition</h3>
<p><strong>Problem</strong>: Ca²⁺ competes with Nd³⁺ for XoxF binding and suppresses XoxF expression</p>
<p><strong>Solution</strong>: Minimize Ca²⁺ in base medium (&lt;10 μM) to favor lanthanide-dependent pathway</p>
<pre><code class="language-python"># Check Ca²⁺ concentration in optimized medium
ca_ingredients = [ing for ing in result['data']['formulation']['ingredients']
                  if 'calcium' in ing['name'].lower()]
if ca_ingredients:
    print(f&quot;WARNING: High Ca²⁺ may suppress XoxF expression&quot;)
</code></pre>
<h3 id="2-methylolanthanin-production">2. Methylolanthanin Production</h3>
<p><strong>Enhancement</strong>: Methylolanthanin enhances Nd uptake by 10-100x</p>
<p><strong>Requirements</strong>:
- Iron limitation (induces siderophore production)
- Proper carbon/nitrogen ratio
- Aerobic conditions</p>
<pre><code class="language-python"># Verify Fe limitation in medium
fe_ingredients = [ing for ing in result['data']['formulation']['ingredients']
                  if 'iron' in ing['name'].lower() or 'Fe' in ing['name']]
# Target: &lt;1 μM Fe to induce methylolanthanin production
</code></pre>
<h3 id="3-ph-and-solubility">3. pH and Solubility</h3>
<p><strong>Issue</strong>: Lanthanides precipitate as hydroxides at pH &gt;8</p>
<p><strong>Solution</strong>: Maintain pH 6.5-7.5, use citrate or acetate buffers (not phosphate - forms insoluble Nd-phosphate)</p>
<h3 id="4-biosorption-vs-bioaccumulation">4. Biosorption vs Bioaccumulation</h3>
<p><strong>Biosorption</strong> (passive, cell surface binding):
- Fast (minutes to hours)
- Reversible
- ~45 mg Nd/g dry biomass (based on Eu data)</p>
<p><strong>Bioaccumulation</strong> (active, intracellular):
- Slower (hours to days)
- Irreversible
- Higher capacity
- Requires metabolically active cells</p>
<p>For maximum Nd depletion, optimize for both.</p>
<h2 id="expected-results">Expected Results</h2>
<h3 id="optimized-mp-medium-formulation-example">Optimized MP Medium Formulation (Example)</h3>
<p>Based on framework data, expect recommendations similar to:</p>
<pre><code>OPTIMIZED MP MEDIUM FOR M. EXTORQUENS AM-1 (1 L)

CARBON &amp; ENERGY:
- Methanol: 5 mL (0.5% v/v)

LANTHANIDE COFACTOR:
- Neodymium chloride (NdCl₃·6H₂O): 2 μM (0.71 mg/L)
  [Adjust based on initial contamination level]

NITROGEN:
- (NH₄)₂SO₄: 1.0 g/L

PHOSPHORUS:
- K₂HPO₄: 0.7 g/L (LOW - avoid Nd precipitation)

SULFUR:
- MgSO₄·7H₂O: 0.2 g/L

TRACE ELEMENTS:
- Fe-EDTA: 0.5 mg/L (LOW - induces methylolanthanin)
- PQQ (pyrroloquinoline quinone): 1 mg/L
- B12 (cyanocobalamin): 0.1 mg/L
- Biotin: 0.02 mg/L
- Thiamine: 0.1 mg/L

CALCIUM (MINIMIZED):
- CaCl₂: &lt;10 μM (CRITICAL: Keep low to favor XoxF)

BUFFER:
- Citrate buffer (pH 7.0): 10 mM
  [NOT phosphate - prevents Nd precipitation]

CONDITIONS:
- pH: 7.0
- Temperature: 30°C
- Agitation: 150 rpm (aerobic)
- Inoculum: 1-2% (mid-log phase)

GROWTH EXPECTATIONS:
- Lag phase: 6-12 h
- Exponential phase: 12-36 h
- Max OD600: 1.5-2.0
- Nd depletion: &gt;80% after 48h (from 50 μM → &lt;10 μM)
</code></pre>
<h3 id="performance-metrics-to-track">Performance Metrics to Track</h3>
<ol>
<li><strong>Growth rate</strong>: OD600 every 4-6 hours</li>
<li><strong>Nd depletion</strong>: ICP-MS or ICP-OES at 0, 24, 48 hours</li>
<li><strong>Cell-associated Nd</strong>: Digest biomass, measure by ICP</li>
<li><strong>XoxF expression</strong>: qRT-PCR or Western blot</li>
<li><strong>Methylolanthanin production</strong>: HPLC-MS (if available)</li>
</ol>
<h2 id="troubleshooting">Troubleshooting</h2>
<h3 id="poor-growth-od600-05-after-48h">Poor Growth (&lt;OD600 0.5 after 48h)</h3>
<p><strong>Check</strong>:
- [ ] PQQ supplementation adequate (0.5-1 mg/L)
- [ ] Methanol concentration (0.5% optimal, toxic &gt;2%)
- [ ] Calcium contamination (&lt;10 μM)
- [ ] pH stability (should stay 6.8-7.2)</p>
<p><strong>Solution</strong>: Run GenomeFunctionAgent auxotrophy detection, add missing vitamins</p>
<h3 id="low-nd-depletion-50-after-48h">Low Nd Depletion (&lt;50% after 48h)</h3>
<p><strong>Check</strong>:
- [ ] Iron concentration (should be &lt;1 μM to induce methylolanthanin)
- [ ] Phosphate concentration (high phosphate precipitates Nd)
- [ ] Cell density (higher biomass = more Nd uptake)
- [ ] pH (Nd precipitates at pH &gt;8)</p>
<p><strong>Solution</strong>: Query bioprocess data for BP-001 (Eu biosorption) to compare conditions</p>
<h3 id="nd-precipitation">Nd Precipitation</h3>
<p><strong>Symptoms</strong>: Cloudy solution, low dissolved Nd despite no uptake</p>
<p><strong>Solution</strong>:
- Lower pH to 6.5
- Switch from phosphate to citrate buffer
- Add 0.1% citrate as Nd chelator (keeps soluble)</p>
<h2 id="cli-alternative">CLI Alternative</h2>
<p>For quick queries without Python scripting:</p>
<pre><code class="language-bash"># Query organism requirements
python -m microgrowagents.agents.genome_function_agent \
  --query &quot;detect auxotrophies&quot; \
  --organism &quot;Methylobacterium extorquens&quot;

# Query lanthanide data
python -m microgrowagents.skills.simple.sheet_query \
  --collection-id cmm \
  --query-type entity_lookup \
  --entity-name neodymium \
  --output-format markdown

# Optimize medium
python -m microgrowagents.skills.workflows.optimize_medium \
  --medium-name &quot;MP medium&quot; \
  --organism &quot;Methylobacterium extorquens AM-1&quot; \
  --optimization-goal growth
</code></pre>
<h2 id="critical-files-referenced">Critical Files Referenced</h2>
<h3 id="agents">Agents:</h3>
<ul>
<li><code>src/microgrowagents/agents/genome_function_agent.py</code> - Organism-specific requirements</li>
<li><code>src/microgrowagents/agents/media_formulation_agent.py</code> - Medium optimization</li>
<li><code>src/microgrowagents/agents/gen_media_conc_agent.py</code> - Concentration predictions</li>
<li><code>src/microgrowagents/agents/media_role_agent.py</code> - Ingredient role classification</li>
</ul>
<h3 id="skills">Skills:</h3>
<ul>
<li><code>src/microgrowagents/skills/simple/sheet_query.py</code> - Query sheets_cmm data</li>
<li><code>src/microgrowagents/skills/workflows/optimize_medium.py</code> - Optimization workflow</li>
<li><code>src/microgrowagents/skills/workflows/recommend_media.py</code> - New medium design</li>
</ul>
<h3 id="data-files">Data Files:</h3>
<ul>
<li><code>data/sheets_cmm/BER_CMM_Data_for_AI_chemicals_extended.tsv</code> - Nd and lanthanide data</li>
<li><code>data/sheets_cmm/BER_CMM_Data_for_AI_genes_and_proteins_extended.tsv</code> - XoxF gene data</li>
<li><code>data/sheets_cmm/BER_CMM_Data_for_AI_bioprocesses_extended.tsv</code> - Biosorption data</li>
<li><code>data/sheets_cmm/publications/</code> - 120+ publications on lanthanide metabolism</li>
<li><code>data/exports/ingredient_properties_20260108.tsv</code> - M. extorquens requirements</li>
</ul>
<h3 id="database">Database:</h3>
<ul>
<li><code>data/processed/microgrow.duckdb</code> - Main database with all query capabilities</li>
</ul>
<h2 id="verification">Verification</h2>
<p>To verify the framework capabilities are loaded:</p>
<pre><code class="language-python"># Check database connection
from pathlib import Path
import duckdb

db_path = Path(&quot;data/processed/microgrow.duckdb&quot;)
assert db_path.exists(), &quot;Database not found - run initialization scripts&quot;

conn = duckdb.connect(str(db_path))

# Check sheets_cmm data loaded
result = conn.execute(&quot;SELECT COUNT(*) FROM sheet_data WHERE table_id LIKE 'cmm_%'&quot;).fetchone()
print(f&quot;CMM entities loaded: {result[0]}&quot;)

# Check publications loaded
result = conn.execute(&quot;SELECT COUNT(*) FROM sheet_publications WHERE collection_id = 'cmm'&quot;).fetchone()
print(f&quot;CMM publications loaded: {result[0]}&quot;)

conn.close()
</code></pre>
<p><strong>Expected output</strong>:
- CMM entities loaded: ~1,763
- CMM publications loaded: ~118</p>
<h2 id="summary">Summary</h2>
<p>To optimize MP medium for M. extorquens AM-1 growth and Nd depletion:</p>
<ol>
<li><strong>Query organism requirements</strong> (GenomeFunctionAgent) → Identify auxotrophies, cofactors</li>
<li><strong>Query lanthanide data</strong> (SheetQuerySkill) → Get optimal Nd concentrations, XoxF data</li>
<li><strong>Optimize formulation</strong> (OptimizeMediumWorkflow) → Generate complete recipe</li>
<li><strong>Verify recommendations</strong> (MediaRoleAgent + literature cross-check)</li>
<li><strong>Export and prepare</strong> → TSV export + preparation protocol</li>
</ol>
<p><strong>Key optimization principles</strong>:
- Minimize Ca²⁺ (&lt;10 μM) to favor XoxF over MxaF
- Induce methylolanthanin (Fe limitation)
- Prevent Nd precipitation (citrate buffer, pH 6.5-7.5)
- Supplement PQQ (0.5-1 mg/L)
- Use 2 μM Nd³⁺ as starting point</p>
<p><strong>Expected outcomes</strong>:
- Growth: OD600 1.5-2.0 in 48h
- Nd depletion: &gt;80% (50 μM → &lt;10 μM in 48h)
- Biosorption capacity: ~40-50 mg Nd/g dry biomass</p>
<p>The framework contains extensive data to support this application - 120+ publications on lanthanide metabolism, optimal concentration data, enzyme systems, and validated biosorption processes.</p>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tabs.link"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>